Reconnect to Redis on connection or time-out error

az-wegift opened the pull request:
Otherwise RedBeat scheduler quits on connection error, e.g. when Redis
is restarted.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sibson/redbeat/80)
<!-- Reviewable:end -->

sibson commented:
This looks like it has less overhead than #78, my concern is that it changes the default behaviour to always reconnect, rather than terminate on failure.  That behaviour should be left in the control of the end-user.

richardasaurus commented:
Perhaps add a configuration option to enable this?

az-wegift commented:
@sibson I have made retrying turned off by default. Tests failures do not seem to be related to changes I've made. I've got the same errors on master branch when I run `tox`. Can you have a quick look, please?

az-wegift commented:
@sibson now tests pass

sibson commented on README.rst:
```diff
@@ -281,8 +281,12 @@ avoid conflicting options. Here follows the example:::
+        'retry_period': 60,
     }
 
+If ``retry_period`` is given, retry connection up to 3 times in ``retry_period`` intervals. If not set, retrying mechanism is not triggered.
```
Why limit to 3 times? Should it also be configurable and/or support infinite retries?

az-wegift commented:
@sibson Added tenacity to requirements. Documentation was incorrect, it is fixed.

sibson commented on redbeat/schedulers.py:
```diff
@@ -39,6 +41,35 @@
+               stop=(stop_after_delay(self.retry_period)),
+               wait=wait_fixed(min(self.retry_period // 2, self.RETRY_MAX_WAIT)))
+        def retrier(*args, **kwargs):
+            method = getattr(self.wrapped_connection, item)
```
Why do we need to call ```getattr``` again here rather than using the previously captured ```method```?

sibson commented on redbeat/schedulers.py:
```diff
@@ -39,6 +41,35 @@
+        @retry(retry=(retry_if_exception_type(redis.exceptions.ConnectionError)
+                      | retry_if_exception_type(redis.exceptions.TimeoutError)),
+               reraise=True,
+               stop=(stop_after_delay(self.retry_period)),
```
Is there a way to never stop?

sibson commented on redbeat/schedulers.py:
```diff
@@ -39,6 +41,35 @@
+                      | retry_if_exception_type(redis.exceptions.TimeoutError)),
+               reraise=True,
+               stop=(stop_after_delay(self.retry_period)),
+               wait=wait_fixed(min(self.retry_period // 2, self.RETRY_MAX_WAIT)))
```
Is there a reason to use wait_fixed rather than wait_exponential or wait_random_exponential?  https://github.com/jd/tenacity#waiting-before-retrying

az-wegift commented:
@sibson thanks for your comments. All fixed.

> Is there a way to never stop?

done

> Is there a reason to use wait_fixed rather than wait_exponential or wait_random_exponential?

changed to exponential

> Why do we need to call getattr again here rather than using the previously captured method?

fixed

There's no additional output while retrying:

```
14:44:18 scheduler.1 | celery beat v4.1.0 (latentcall) is starting.
14:44:18 scheduler.1 | __    -    ... __   -        _
14:44:18 scheduler.1 | LocalTime -> 2018-08-23 14:44:18
14:44:18 scheduler.1 | Configuration ->
14:44:18 scheduler.1 |     . broker -> amqp://guest:**@localhost:5672//
14:44:18 scheduler.1 |     . loader -> celery.loaders.app.AppLoader
14:44:18 scheduler.1 |     . scheduler -> redbeat.schedulers.RedBeatScheduler
14:44:18 scheduler.1 |        . redis -> redis-sentinel://redis-sentinel:26379/0
14:44:18 scheduler.1 |        . lock -> `redbeat::lock` 25.00 minutes (1500s)
14:44:18 scheduler.1 |     . logfile -> [stderr]@%INFO
14:44:18 scheduler.1 |     . maxinterval -> 5.00 minutes (300s)
14:44:18 scheduler.1 | [2018-08-23 14:44:18,705: INFO/MainProcess] Initialising Celery task error reporting for logger root
14:44:18 scheduler.1 | [2018-08-23 14:44:18,706: INFO/MainProcess] Initialising Celery task error reporting for logger celery.task
14:44:18 scheduler.1 | [2018-08-23 14:44:18,706: INFO/MainProcess] beat: Starting...
```

It would be possible to add warning message, but I don't believe it's necessary and will certainly complicate code.

az-wegift commented:
> Is there a reason to use wait_fixed rather than wait_exponential or wait_random_exponential?

I don't think that makes a difference, as I'd assume redis is on local network and there's no harm trying to connect to it frequently. But given using exponential looks simpler (no need for this `min()` function), I've changed it.

sibson commented:
Regarding, logging redis connection state, if I'm trying to debug why tasks aren't being triggered and its because redbeat is configured to run against the wrong server or the server is unreachable due to networking issues, how would I identify that? In my mind I'd look for a log message indicating the status of the connection or that reties are happening. It seems like [tenacity supports that](https://github.com/jd/tenacity#before-and-after-retry-and-logging) reasonably easily.   I'm open to other options but I think there needs to be a reasonable story for end users.

Other than the debugging connection issues question I think this looks good.

az-wegift commented:
@sibson Thanks for the tip. It's been added in the newer version! I will add this.

az-wegift commented:
@sibson added. Let me test entire thing again in our staging environment for a while before merging. Just let me know if it's in the shape to be merged.

sibson commented:
lgtm, let me know when you are feel its ready to be merged.

az-wegift commented:
@sibson It was running for us without issues for this time. Please feel free to merge. Could you make a release on PyPI, please?

az-wegift commented on setup.py:
```diff
@@ -32,7 +32,8 @@
         'celery',
-        'python-dateutil'
+        'python-dateutil',
+        'tenacity'
```
I don't think it's worth spending time on making this conditional as tenacity is a small library that would not add any significant overhead, but please feel free to propose a change.

az-wegift commented:
@sibson any plans on making a release on PyPI?

sibson commented:
done 0.12.0, thanks for the nudge.

az-wegift commented:
Thanks!
